zio-http/src/main/scala/zio/http/DnsResolver.scala

```diff
@@ -0,0 +1,285 @@
+      maxCount: Int,
+      maxConcurrentResolutions: Int,
+      expireAction: ExpireAction,
+      refreshRate: Duration,
+    ): ZIO[Scope, Nothing, DnsResolver] =
+      for {
+        semaphore <- Semaphore.make(maxConcurrentResolutions)
+        entries   <- Ref.make(Map.empty[String, CacheEntry])
+        cachingResolver = new CachingResolver(resolver, ttl, unknownHostTtl, maxCount, expireAction, semaphore, entries)
+        _ <- cachingResolver.refreshAndCleanup().scheduleFork(Schedule.fixed(refreshRate))
+      } yield cachingResolver
+  }
+
+  private[http] def snapshot(): ZIO[DnsResolver, Nothing, Map[String, CacheEntry]] =
+    ZIO.service[DnsResolver].flatMap {
+      case cachingResolver: CachingResolver => cachingResolver.snapshot()
+      case _ => ZIO.dieMessage(s"Unexpected DnsResolver implementation: ${getClass.getName}")
+    }
+
+  final case class DnsResolverConfig(
+    ttl: Duration,
+    unknownHostTtl: Duration,
+    maxCount: Int,
+    maxConcurrentResolutions: Int,
+    expireAction: ExpireAction,
+    refreshRate: Duration,
+  )
+
+  object DnsResolverConfig {
+    val default: DnsResolverConfig = DnsResolverConfig(
+      ttl = 10.minutes,
+      unknownHostTtl = 1.minute,
+      maxCount = 4096,
+      maxConcurrentResolutions = 16,
+      expireAction = ExpireAction.Refresh,
+      refreshRate = 2.seconds,
+    )
+
+    val config: Config[DnsResolverConfig] =
+      (Config.duration("ttl").withDefault(default.ttl) ++
+        Config.duration("unknown-host-ttl").withDefault(default.unknownHostTtl) ++
+        Config.int("max-count").withDefault(default.maxCount) ++
+        Config.int("max-concurrent-resolutions").withDefault(default.maxConcurrentResolutions) ++
+        ExpireAction.config.nested("expire-action").withDefault(default.expireAction) ++
+        Config.duration("refresh-rate").withDefault(default.refreshRate)).map {
+        case (ttl, unknownHostTtl, maxCount, maxConcurrentResolutions, expireAction, refreshRate) =>
+          DnsResolverConfig(ttl, unknownHostTtl, maxCount, maxConcurrentResolutions, expireAction, refreshRate)
+      }
+  }
+
+  def cached(
+    ttl: Duration = 10.minutes,
+    unknownHostTtl: Duration = 1.minute,
+    maxCount: Int = 4096,
+    maxConcurrentResolutions: Int = 16,
+    expireAction: ExpireAction = ExpireAction.Refresh,
+    refreshRate: Duration = 2.seconds,
+    implementation: DnsResolver = SystemResolver(),
+  ): ZLayer[Any, Nothing, DnsResolver] =
+    ZLayer.scoped {
+      CachingResolver
+        .make(
+          implementation,
+          ttl,
+          unknownHostTtl,
+          maxCount,
+          maxConcurrentResolutions,
+          expireAction,
+          refreshRate,
+        )
+    }
+
+  val configured: ZLayer[Any, Config.Error, DnsResolver] =
+    ZLayer(ZIO.config(DnsResolverConfig.config.nested("zio.http.dns"))) >>> fromConfig
+
+  val default: ZLayer[Any, Nothing, DnsResolver] = ZLayer.succeed(DnsResolverConfig.default) >>> fromConfig
```

Comment:
Is there a good rerason to have both `configured` and `default`? Wouldn't a user always want `configured` (so we could rename that to `default`)?